### PR TITLE
Open zeplin

### PR DESCRIPTION
**Pull request by Samson303**
Changes: 

- Moved game logic out of token contract
- token contract inherits from OpenZeplin ERC20 and Ownable
- Moved all to Rinkeby (all contract details ex. address in readMe)

### CODE REVIEW

File: contracts/MetaxuGameAdmin.sol [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+import './MTX.sol';
+import 'node_modules/@openzeppelin/contracts/access/Ownable.sol';
+
+
+contract MetaxuGameAdmin is Ownable {
+    MTX private _mtx;
+    mapping (address => bool) recievedAirDrop_;
+    mapping (address => bool) eligableForAirDrop_;
+    mapping(bytes32 => uint256)easterEggToReward_;
```
Comment (diminator) [inline contracts/MetaxuGameAdmin.sol:11]: style nit: spaces after `)`

File: contracts/MetaxuGameAdmin.sol [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        setToken(mtx);
+        isAdmin_[msg.sender] = true;
+    }
+
+    function setToken(MTX token) public onlyOwner {
+            _mtx = token;
+        }
+
+    function getSha256(string memory str) public pure returns (bytes32) {
+        bytes32 hash = sha256(abi.encodePacked(str));
```
Comment (diminator) [inline contracts/MetaxuGameAdmin.sol:32]: can also use `keccak`. is more secure (SHA3). but maybe costs more gas...

File: contracts/MetaxuGameAdmin.sol [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        return hash;
+        }
+
+
+    function AddToListForAirDrop(address[] memory listOfRecepients) public returns (bool success) {
+        require(isAdmin_[msg.sender], 'must be admin');
+
+        for (uint256 index = 0; index < listOfRecepients.length; index++) {
+            if (eligableForAirDrop_[listOfRecepients[index]]) {
```
Comment (diminator) [inline contracts/MetaxuGameAdmin.sol:41]: could be removed as you can always set `true` without check

File: contracts/MetaxuGameAdmin.sol [hunk 4]
```diff
@@ -0,0 +1,89 @@
+pragma solidity >=0.4.17 <0.8.0;
+
+import './MTX.sol';
+import 'node_modules/@openzeppelin/contracts/access/Ownable.sol';
+
+
+contract MetaxuGameAdmin is Ownable {
+    MTX private _mtx;
+    mapping (address => bool) recievedAirDrop_;
```
Comment (diminator) [inline contracts/MetaxuGameAdmin.sol:9]: `receive`, not `recieve`

**Review by diminator — APPROVED** [review]
very clean

File: contracts/MTX.sol [hunk 5]
```diff
@@ -1,97 +1,36 @@
 pragma solidity >=0.4.17 <0.8.0;
 
 import 'node_modules/@openzeppelin/contracts/token/ERC20/ERC20.sol';
+import 'node_modules/@openzeppelin/contracts/access/Ownable.sol';
 
 
-
-contract MTX is ERC20 {
+contract MTX is ERC20, Ownable {
```
Comment (diminator) [inline contracts/MTX.sol:7]: for alternatives:
https://github.com/OpenZeppelin/openzeppelin-contracts/tree/master/contracts/token/ERC20
I could see `Pausable` as a nice feature

File: contracts/MTX.sol [hunk 6]
```diff
@@ -1,97 +1,36 @@
...
-    event AirDropped(address indexed receiver, uint tokens);
-    event EasterEggHasBeenRedeemed(address redeemer, uint tokensAwarded);
-    event AdminAdded(address approver, address newAdmin);
-
-
-    function getSha256(string memory str) public pure returns (bytes32) {
-        bytes32 hash = sha256(abi.encodePacked(str));
-        return hash;
-        }
+    function mint (address callingContract, address accountMintedTo, uint256 amount) public  {
```
Comment (diminator) [inline contracts/MTX.sol:21]: there's also a ERC20 templates that have `Minter` roles. maybe need to investigate that

Comment (diminator) [reply contracts/MTX.sol:21]: https://docs.openzeppelin.com/contracts/3.x/api/presets#ERC20PresetMinterPauser

File: contracts/MetaxuGameAdmin.sol [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        return 'success';
+    }
+
+
+    function redeemEasterEgg(string memory attemptedSecret) public returns (string memory success) {
+        bytes32 hashOfAttempt = getSha256(attemptedSecret);
+        uint256 reward = easterEggToReward_[hashOfAttempt];
+
+        require(reward > 0, 'wrong input');
+        require(!hasBeenRedeemed_[hashOfAttempt], 'easter egg has been redeemed');
```
Comment (diminator) [inline contracts/MetaxuGameAdmin.sol:78]: this mean we'll only can use one secret each time and never reuse it.
If we want to reuse secrets we could do something like
```
easterEggToReward_[hashOfAttempt] = 0
```
at the end and remove `hasBeenRedeemed`

**Review by diminator — CHANGES_REQUESTED** [review]

File: contracts/MetaxuGameAdmin.sol (hunk 7 above)
Comment (Samson303) [reply contracts/MetaxuGameAdmin.sol]: has been removed and implemented a secretData struct which holds property of amountLeft that can be topped up

File: contracts/MTX.sol (hunk 6 above)
Comment (Samson303) [reply contracts/MTX.sol]: implemented and minters will be msg.sender and game contract (added after uploading of contract)

File: contracts/MTX.sol (hunk 5 above)
Comment (Samson303) [reply contracts/MTX.sol]: using openZeplin ERC20PresetMinterPauser

File: contracts/MetaxuGameAdmin.sol (hunk 2 above)
Comment (Samson303) [reply contracts/MetaxuGameAdmin.sol]: changed to keccak256 on new commit

File: contracts/MetaxuGameAdmin.sol (hunk 1 above)
Comment (Samson303) [reply contracts/MetaxuGameAdmin.sol]: changed on new commit

**Review by diminator — APPROVED** [review]

**Review by diminator — APPROVED** [review]

**Pull request merged**